vendor/engines/secure_rooms/app/services/secure_rooms/scan_authorizer.rb

```diff
@@ -0,0 +1,85 @@
+module SecureRooms
+
+  class ScanAuthorizer
+
+    include TextHelpers::Translation
+
+    def initialize(card_number, reader_identifier, controller_identifier)
+      @user = User.find_by(card_number: card_number)
+      @card_reader = SecureRooms::CardReader.find_by(
+        card_reader_number: reader_identifier,
+        control_device_number: controller_identifier,
+      )
+    end
+
+    def perform
+      if error_reasons.present?
+        set_error_response
+      elsif denial_reasons.present?
+        set_denial_response
+      else
+        set_accounts_response
+      end
+    end
+
+    def response
+      @response ||= { json: @response_json, status: @response_status }
+    end
+
+    def user_accounts
+      @user_accounts ||= @user.accounts_for_product(@card_reader.secure_room)
+    end
+
+    def set_accounts_response
+      @response_status = user_accounts.many? ? :multiple_choices : :ok
+      @response_json = {
+        response: "select_account",
+        # TODO: needs a legitimage tablet_identifier once tablet exists
+        tablet_identifier: "abc123",
+        name: @user.full_name,
+        accounts: SecureRooms::AccountPresenter.wrap(user_accounts),
+      }
+    end
+
+    def set_error_response
+      @response_status = :not_found
+      @response_json = {
+        response: "deny",
+        reason: error_reasons,
+      }
+    end
+
+    def set_denial_response
+      @response_status = :forbidden
+      @response_json = {
+        response: "deny",
+        reason: denial_reasons,
+      }
+    end
+
+    def denial_reasons
+      [].tap do |reasons|
+        reasons << text("errors.no_accounts") if user_accounts.empty?
```

Comment:
Think we're actually going to want something that's more as soon as you get a denial return. I don't think we need to check and return every problem. Only the first.